File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = lower_bound(nums, target);
        int r = upper_bound(nums, target);
        return {l, r};
    }

private:
    int lower_bound(vector<int> &nums, int target){
        int l = 0, r = nums.size()-1;
        while(l <= r){
            int mid = midpoint(l, r);
            if (nums[mid] == target){
                if (mid == 0) return mid;
                else {
                    if (nums[mid-1] != target) return mid;
                    else r = mid - 1;
                }
            } else if (nums[mid] > target){
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return -1;
    }
    int upper_bound(vector<int> &nums, int target){
        int l = 0, r = nums.size()-1;
        while(l <= r){
            int mid = midpoint(l, r);
            if (nums[mid] == target){
                if (mid == nums.size()-1) return mid;
                else {
                    if (nums[mid+1] != target) return mid;
                    else l = mid + 1;
                }
            } else if (nums[mid] > target){
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return -1;
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.go
func searchRange(nums []int, target int) []int {
    l := lower_bound(nums, target)
    r := upper_bound(nums, target)

    return []int{l, r}
}

func lower_bound(nums []int, target int) int {
    l, r := 0, len(nums)-1
    for l <= r {
        mid := l + (r - l) / 2
        
        if nums[mid] == target {
            if mid == 0 || nums[mid] != nums[mid-1] {
                return mid
            } else {
                r = mid - 1
            }
        } else if nums[mid] > target {
            r = mid - 1
        } else {
            l = mid + 1
        }
    }
    return -1
}

func upper_bound(nums []int, target int) int {
    l, r := 0, len(nums)-1
    for l <= r {
        mid := l + (r - l) / 2
        
        if nums[mid] == target {
            if mid == len(nums)-1 || nums[mid] != nums[mid+1] {
                return mid
            } else {
                l = mid + 1
            }
        } else if nums[mid] > target {
            r = mid - 1
        } else {
            l = mid + 1
        }
    }
    return -1
}
